servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherBuffer.java

```diff
@@ -249,57 +290,111 @@ public void onComplete() {
                 }
                 if (cMaybeAccumulator == null) {
                     // This is the first received accumulator (first boundary start); so we just store the accumulator
-                    // to accumulate and request items from itemsSubscription.
-                    if (maybeAccumulatorUpdater.compareAndSet(this, null, nextAccumulator)) {
-                        itemsSubscription.request(firstItemsRequestN);
-                        bSubscription.request(1); // since we did not emit the accumulator
+                    // to accumulate and request items from tSubscription.
+                    if (maybeAccumulatorUpdater.compareAndSet(this, null, toCounting(nextAccumulator))) {
+                        tSubscription.request(firstItemsRequestN);
+                        // Since we did not emit the accumulator we request one more to observe the next boundary:
+                        bSubscription.request(1);
                         return;
                     }
                 } else if (cMaybeAccumulator == ADDING) {
                     // Hand-off emission of nextAccumulator to the thread that is ADDING
-                    if (maybeAccumulatorUpdater.compareAndSet(this, ADDING, nextAccumulator)) {
-                        bSubscription.request(1); // since we did not emit the accumulator
-                        return;
-                    }
-                } else if (cMaybeAccumulator instanceof ItemsTerminated) {
-                    bSubscription.cancel();
-                    if (maybeAccumulatorUpdater.compareAndSet(this, cMaybeAccumulator, TERMINATED)) {
-                        @SuppressWarnings("unchecked")
-                        ItemsTerminated<T, B> itemsTerminated = (ItemsTerminated<T, B>) cMaybeAccumulator;
-                        terminateTarget(itemsTerminated.accumulator, target, itemsTerminated.terminalNotification,
-                                bSubscription);
+                    if (maybeAccumulatorUpdater.compareAndSet(this, ADDING,
+                            new NextAccumulatorHolder<>(nextAccumulator))) {
+                        // Thread that is ADDING will observe the change and emit the current accumulator
                         return;
                     }
+                } else if (NextAccumulatorHolder.class.equals(cMaybeAccumulator.getClass())) {
+                    // Hand-off emission of one more nextAccumulator to the thread that is still ADDING, discarding it.
```

Comment:
Similarly here about discarding emitted boundaries. I am not saying we should fix it, but I believe at the very least we should include some warning in the javadoc, maybe: 
"boundaries should *not* be used as lifecycle aware objects, `finish` is not guaranteed to be called."

Reply:
Clarified in javadoc.